design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+    Future<Standalone<StringRef>> read(location start, location end, CheckHashes ch);
+    // ...
+};
+```
+
+Internally, the DiskQueue adds page headers every 4K, which are stripped out
+from the returned data.  Therefore, the length of the result will not be the
+same as `end-start`, intentionally.  For this reason, the API is `(start, end)`
+and not `(start, length)`.
+
+Spilled data, when using spill-by-value, was resistent to bitrot via data being
+checksummed interally within SQLite's B-tree.  Now that reads can be done
+directly, the responsibility for verifing data integrity falls upon the
+DiskQueue.  `CheckHashes::YES` will cause the DiskQueue to use the checksum in
+each DiskQueue page to verify data integrity.  If an externally maintained
+checksums exists to verify the returned data, then `CheckHashes::NO` can be
+used to elide the checksumming.
+
+### Popping
+
+As storage servers persist data, they send `pop(tag, version)` requests to the
+transaction log to notify it that it is allowed to discard data for `tag` up
+through `version`.  Once all the tags have been popped from the oldest commit
+in the DiskQueue, the tail of the DiskQueue can be discarded to reclaim space.
+
+
+Each time FoundationDB goes through a recovery, it will recruit a new
+generation of transaction logs.  This new generation of transaction logs will
+often be recruited on the same worker that hosted the previous generation's
+transaction log.  The old generation of transaction logs will only shut down
+once all the data that they have has been fully popped.  This means that there
+can be multiple instances of a transaction log 
+
+*********************************************************
+*                   SharedTLog                          *
+*                                                       *
+* +--------+--------+--------+--------+--------+        *
+* | TLog 1 | TLog 2 | TLog 3 | TLog 4 | TLog 5 |        *
+* +--------+--------+--------+--------+--------+        *
+*   ^ popping         ^spilling         ^committing     * 
+*********************************************************
+
+
+
+
+### Transaction State Store
+
+For FDB to perform a recovery, there is information that it needs to know about
+the database, such as the configuration, worker exclusions, backup status, etc.
+These values are stored into the database in the `\xff` system keyspace.
+However, during a recovery, FDB can't read this data from the storage servers,
+because recovery hasn't completed, so it doesn't know who the storage servers
+are yet.  Thus, a copy of this data is held in-memory on every proxy in the
+*transaction state store*, and durably persisted as a part of commits on the
+transaction logs.  Being durably stored on the transaction logs means the list
+of transaction logs can be fetched from the coordinators, and then used to load
+the rest of the information about the database.
+
+The in-memory storage engine writes an equal amount of mutations and snapshot
```

Comment:
I'm a little confused by this paragraph. Are you saying that the implementation of the in-memory engine works by having the transaction logs constantly being inserted with approximately an equal amount of new data mutations as well as snapshot entries that in combination create a "virtual snapshot" that can be regenerated in-memory by reading the queue?